R-package/R/xgboost.R

```diff
@@ -512,6 +516,9 @@ check.can.use.qdm <- function(x, params) {
       return(FALSE)
     }
   }
+  if (NROW(eval_set)) {
```

Comment:
What does this imply? If `eval_set` has a valid number of rows then we can't use qdm?

Reply:
Yes, because it then slices the DMatrix that gets created.

Reply:
Does the slicing need to happen after the DMatrix is created?

Reply:
Yes, because otherwise there'd be issues with things like needing to make sure categorical 'y' and features have the same encodings between the two sets, objects from package `Matrix` returning a different class when they are sliced, and so on.

Reply:
Hmm, so from your perspective the DMatrix is more suitable for slicing than built-in classes...

Reply:
Unfortunately yes.

Reply:
That's indeed unfortunate.

Reply:
I have a strong preference for not using the DMatrix slice method. The DMatrix can be sliced only because it's a CSR for the simplest case. But it's considered an internal cache for XGBoost instead of something that can be manipulated from outside. For instance, with external memory, the DMatrix actually represents an async fetcher of a disk file.

Reply:
Well, the alternative would be to:
* Restructure the code such that 'x' and 'y' get processed earlier.
* Make it so that characters in 'x' get converted to factor before passing them to `xgb.DMatrix`, and so that `data.frames` get their subclasses (like `data.table`) removed beforehand.
* Add a custom slicer for `Matrix` classes. Either that, or additional castings of storage format, or pull an extra dependency for efficient slicing.

But it'd end up being inefficient either way. The moreso considering that on the R side, the slicing would happen with a vector of random indices on one of the following:
* A column-major matrix of 8-byte elements (likely slower than a CSR).
* A list of vectors (what a dataframe is behind the scenes).
* A CSC matrix, which first will get converted to that format from CSR (`Matrix` doesn't slice CSR directly), and the slice getting converted to COO.

It could in theory be more efficient to do the slicing in R for base `matrix` classes, but probably not so much for the others.

Reply:
Any suggestion for the future implementation of CV optimization previously mentioned? It's designed for the the Qdm and the external memory version of Qdm.

Reply:
Sounds good and quite helpful for the CV function indeed. But I don't think it's very relevant here, since unlike `xgb.cv`, (a) it's only doing two slicing operations and only one quantization/binning, (b) it doesn't accept an arbitrary `xgb.DMatrix` - instead, it creates it internally from a small subset of allowed classes.

Hence, it doesn't need to consider special cases like external memory or distributed mode, and there isn't too much room for improvement in terms of speed savings.

Reply:
Makes sense. We will have a second cv function in the future for high-level inputs (like  data.table and iterator etc).